### PR TITLE
ledger-on-sql: Make sure the ledger picks up where it left off.

### LINKED ISSUE #4223: Ensure that the Dispatcher head is correct on startup in the KV SQL Ledger
Currently, the KV-based SQL ledger always starts dispatching from index 1 (the start), rather than wherever it was left off. This was copied from the in-memory ledger, where this trivially works because it always starts with no data.

We need to get this information from the database.

In addition, I think we need a conformance test (in the Ledger API Test Tool) to ensure that everything continues working after restarting the ledger. This will have to be run separately to all other tests, as it needs control over the process and will break anything running concurrently.

### PR DESCRIPTION
**Pull request by SamirTalwar**
This means seeding the dispatcher with the correct head, rather than starting at 1 every time.

I'd like to move the test case into `ParticipantStateIntegrationSpecBase`, but it's going to require more thought around acquiring and releasing the participant state resource.

Closes #4223.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/package.scala [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.on
```
Comment (stefanobaghino-da) [inline ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/package.scala]: `on` is a quite generic term. From a cursory look it doesn't appear to be a reserved keyword in neither Java nor Scala, so I guess it's unlikely to be an issue, but I would consider dropping it or replacing it. `com.daml.ledger.sql` seems equally telling.

File: ledger/ledger-on-sql/src/test/suite/scala/com/daml/ledger/on/sql/RestartSpec.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class RestartSpec
+    extends AsyncWordSpec
+    with Matchers
+    with BeforeAndAfterEach
+    with AkkaBeforeAndAfterAll {
+
+  private val root = Files.createTempDirectory(getClass.getSimpleName)
+
+  private def start(id: String): Resource[KeyValueParticipantState] = {
+    newLoggingContext { implicit logCtx =>
```
Comment (stefanobaghino-da) [inline ledger/ledger-on-sql/src/test/suite/scala/com/daml/ledger/on/sql/RestartSpec.scala]: if you want you can also use `newLoggingContext` after the `in` method when defining tests. Whatever you feel makes more sense.

File: ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/package.scala (hunk 1 above)
Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/package.scala]: That is fair; happy to rename this along with renaming the Bazel package at a later point (after we decide as a group).

File: ledger/ledger-on-sql/src/test/suite/scala/com/daml/ledger/on/sql/RestartSpec.scala (hunk 2 above)
Comment (SamirTalwar) [reply ledger/ledger-on-sql/src/test/suite/scala/com/daml/ledger/on/sql/RestartSpec.scala]: I'll leave it here for now but I'll consider it when this gets refactored into `ParticipantStateIntegrationSpecBase`.

**Pull request merged**